src/launch/snp.rs

```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! An implementation of the SEV-SNP launch process as a type-state machine.
+//! This ensures (at compile time) that the right steps are called in the
+//! right order.
+
+pub use crate::launch::{
+    SnpFinish, SnpPageType, SnpPolicy, SnpPolicyFlags, SnpStart, SnpUpdate, VmplPerms,
+};
+
+use crate::kvm::types::*;
+use crate::launch::linux::ioctl::*;
+
+use std::io::Result;
+use std::marker::PhantomData;
+use std::os::unix::io::AsRawFd;
+
+/// Launcher type-state that indicates a brand new launch.
+pub struct New;
+
+/// Launcher type-state that indicates a SNP in-progress.
+pub struct Started;
+
+/// Facilitates the correct execution of the SEV launch process.
+pub struct Launcher<'a, T, U: AsRawFd, V: AsRawFd> {
+    vm_fd: &'a mut U,
+    sev: &'a mut V,
+    state: PhantomData<T>,
+}
+
+impl<'a, T, U: AsRawFd, V: AsRawFd> Launcher<'a, T, U, V> {
+    /// Give access to the vm fd to create vCPUs or such.
+    pub fn as_mut_vmfd(&mut self) -> &mut U {
+        self.vm_fd
+    }
+}
```

Comment:
Why not just `impl AsRawFd`? The `Launcher` itself is handling all the relevant ioctls on `/dev/sev`. So it doesn't make any sense to get the fd for `/dev/sev` out of the `Launcher` type. Right?

Reply:
If I input a real kvm `VmFd`, I can use the methods of it then. Otherwise the VmFd is totally blocked by the Launcher holding a mutable reference.

Reply:
Then `impl AsRef` and `impl AsMut`.

Reply:
yeah, maybe... I think I wanted to keep the door to the mut sev fd open

Reply:
I'm of the opposite opinion: nothing should have mut access to the sev fd while we're launching this guest.

Reply:
Well, we need to create vcpus after SNP_LAUNCH_INIT and before SNP_LAUNCH_FINISH, and this can only be done with a VmFd.

Reply:
Yes, I agree. I'm not sure how that affects my point. Launch should `impl AsRef<VmFd>` but not `impl AsRef<SevFd>`.

Reply:
Ah, right... so let's implement `impl AsRef<U>` and `impl AsMut<VmFd>`

Reply:
done